keyspace/key_space.go

```diff
@@ -184,24 +190,32 @@ func (ks *KeySpace) Watch(ctx context.Context, client clientv3.Watcher) error {
 				}
 			} else if err != nil {
 				return err // All other errors are fatal.
-			} else if resp.Header.Revision < nextRevision {
-				// TODO(johnny): Extra logging to better understand root cause(s) of issue #248.
+			} else if len(resp.Events) > 0 {
+				// Events are ordered by ModRevision, so the first is the min, and last is max
+				var minRevision = resp.Events[0].Kv.ModRevision
+				resumeRevision = resp.Events[len(resp.Events)-1].Kv.ModRevision
 				log.WithFields(log.Fields{
-					"header":           resp.Header,
-					"isProgressNotify": resp.IsProgressNotify(),
-					"cancelled":        resp.Canceled,
-					"created":          resp.Created,
-					"compactRevision":  resp.CompactRevision,
-					"numEvents":        len(resp.Events),
-					"watchIter":        attempt,
-				}).Warn("received duplicate Etcd watch revision (ignoring)")
-			} else {
+					"header":          resp.Header,
```

Comment:
optional supernit:

this is starting to get into "preference", but it's grounded enough to share, so I'm doing so and do with it what you will:

In your shoes, I would inline the lookups that yield `minModRevision` and `maxModRevision` in the log statement.
Then, below, I would have resumeRevision separately be updated with a comment explaining why we're setting it to the maximum mod revision. (I'd also comment in the `else` block below, as to why we're using `resp.Header.Revision`.

This does a couple of things. First, it makes the log statement side-effect-less and easy to move or delete, with no dependence on other stack variables. Second, it keeps separation between SLOC which are "about debug logging" from lines that are actual algorithm, which I find easier as a reader. Right now, it's not immediately obvious what role `minRevision` and `resumeRevision` are playing in the algorithm. It's not clear that `minRevision` is _only_ for logging out, and conversely the update of `resumeRevision` has _great_ significance.